### DISCUSSION TITLE
How do I craft a PSBT with OP_RETURN outputs?

### DISCUSSION THREAD

**Discussion by Zodomo** [paulmillr/micro-btc-signer#83] [Q&A]
I am trying to figure out how to craft a PSBT with 3-4 OP_RETURN outputs to store metadata within a Bitcoin transaction and issue a payment to a recipient address. It appears this utility is able to do this, but I am not very experienced with Bitcoin Script so I am not sure if I am doing this correctly.

Based on my own experimentation, does the following workflow make sense? For the purpose of this discussion, the OP_RETURN data is short. I am aware of the 80 byte limit per OP_RETURN, but have chosen to not maximally optimize their packing for this example.

```ts
import * as btc from '@scure/btc-signer';

// Create a new transaction
const tx = new btc.Transaction({
    allowUnknownOutputs: true, // Allows OP_RETURN outputs
  });

// Add the OP_RETURN outputs
tx.addOutput({
  script: btc.Script.encode([btc.OP.RETURN, Buffer.from('Hello, World!')]),
  amount: 0n,
});

tx.addOutput({
  script: btc.Script.encode([btc.OP.RETURN, Buffer.from('This is the second OP_RETURN output')]),
  amount: 0n,
});

tx.addOutput({
  script: btc.Script.encode([btc.OP.RETURN, Buffer.from('And this is the third one')]),
  amount: 0n,
});

// Add a payment output to a P2TR address
const recipientAddress = 'bc1p5cyxnuxmeuwuvkwfem96lqzszd02n6xdcjrs20cac6yqjjwudpxqkedrcr';
const paymentAmount = btc.Decimal.decode('0.001'); // 0.001 BTC

// Create a P2TR output
tx.addOutputAddress(recipientAddress, paymentAmount);

// Encode the transaction as a PSBT
const psbt = tx.toPSBT();
console.log(Buffer.from(psbt).toString('base64'));
```

I assume this PSBT would then be submitted to a wallet for signing and transmission. However, I am not sure how to go about including inputs. Do I retrieve large enough UTXO(s) from the connected user's Bitcoin address to cover the transaction?

**Comment by paulmillr** [comment]
This is being investigated. Perhaps a readme entry will be added soon.

**Comment by paulmillr** [comment]
Ordinals got easier with 1.3.0 https://github.com/paulmillr/scure-btc-signer/releases/tag/1.3.0

**Reply by Zodomo** [reply]
Thanks for this update, however, I was asking as a means of just including data alongside a transaction. I am uninterested in the tradable nature of ordinals for my purposes. I just want to be able to bundle some user data alongside a proposed BTC TX, and figured storing it directly as OP_RETURN data that I could then reindex later on if needed would be a solid solution. Is there a way to use scure-btc-signer to craft transactions that also have OP_RETURN outputs (not ordinals via embedding in witness data)?

**Reply by paulmillr** [reply]
inscriptions on top of ordinals is the simplest and standard way of adding random data to btc blockchain. I think it would work great for your purpose.

you could invent something separate but why? when you can use a standard

**Reply by Zodomo** [reply]
Ordinals are tradable. I'm just looking to store purchase data alongside transactions. I don't want that purchase data to be tradable, nor do I want my users seeing useless (to them) purchase records in their Ordinal wallets. I'm trying to use Bitcoin for this data instead of a database, and intend to use OP_RETURN outputs so I can index user purchases if needed. I get that it's nonstandard, but I think that's fine given it's only for record-keeping purposes.
